adapter/memQuota/memQuota.go

```diff
@@ -0,0 +1,296 @@
+	windows map[string]*rollingWindow
+
+	// two ping-ponging maps of active dedup ids
+	recentDedup map[string]int64
+	oldDedup    map[string]int64
+
+	// used for reaping dedup ids
+	ticker *time.Ticker
+
+	// indirection to support fast deterministic tests
+	getTick func() int32
+
+	logger adapter.Logger
+}
+
+// we maintain a pool of these for use by the makeKey function
+type keyWorkspace struct {
+	keys   []string
+	buffer []byte
+}
+
+// pool of reusable keyWorkspace structs
+var keyWorkspacePool = sync.Pool{New: func() interface{} { return &keyWorkspace{} }}
+
+var (
+	name = "istio/memQuota"
+	desc = "Simple volatile memory-based quotas."
+	conf = &config.Params{MinDeduplicationWindowSeconds: 1}
+)
+
+const (
+	ticksPerSecond = 10
+)
+
+// Register records the builders exposed by this adapter.
+func Register(r adapter.Registrar) {
+	r.RegisterQuota(newBuilder())
+}
+
+func newBuilder() builder {
+	return builder{adapter.NewDefaultBuilder(name, desc, conf)}
+}
+
+func (builder) NewQuota(env adapter.Env, c adapter.AspectConfig, d map[string]*adapter.QuotaDefinition) (adapter.QuotaAspect, error) {
+	return newAspect(env, c.(*config.Params), d)
+}
+
+// newAspect returns a new aspect.
+func newAspect(env adapter.Env, c *config.Params, definitions map[string]*adapter.QuotaDefinition) (adapter.QuotaAspect, error) {
+	mq := &memQuota{
+		definitions: definitions,
+		cells:       make(map[string]int64),
+		windows:     make(map[string]*rollingWindow),
+		recentDedup: make(map[string]int64),
+		oldDedup:    make(map[string]int64),
+		ticker:      time.NewTicker(time.Duration(c.MinDeduplicationWindowSeconds) * time.Second),
+		getTick:     func() int32 { return int32(time.Now().UnixNano() / (1000000000 / ticksPerSecond)) },
+		logger:      env.Logger(),
+	}
+
+	go func() {
+		for range mq.ticker.C {
+			mq.reapDedup()
+		}
+	}()
+
+	return mq, nil
+}
+
+func (mq *memQuota) Close() error {
+	mq.ticker.Stop()
+	return nil
+}
+
+func (mq *memQuota) Alloc(args adapter.QuotaArgs) (int64, error) {
+	return mq.alloc(args, false)
+}
+
+func (mq *memQuota) AllocBestEffort(args adapter.QuotaArgs) (int64, error) {
+	return mq.alloc(args, true)
+}
+
+func (mq *memQuota) alloc(args adapter.QuotaArgs, bestEffort bool) (int64, error) {
+	return mq.common(args, func(d *adapter.QuotaDefinition, key string) int64 {
+		result := args.QuotaAmount
+
+		// we optimize storage for non-expiring quotas
+		if d.ExpirationSeconds == 0 {
+			inUse := mq.cells[key]
+
+			if result > d.MaxAmount-inUse {
+				if bestEffort {
+					// grab as much as we can
+					result = d.MaxAmount - inUse
+				} else {
+					result = 0
+				}
+			}
+
+			if result > 0 {
+				mq.cells[key] = inUse + result
+			}
+		} else {
+			window, ok := mq.windows[key]
+			if !ok {
+				window = newRollingWindow(d.MaxAmount, d.ExpirationSeconds*ticksPerSecond)
+				mq.windows[key] = window
+			}
+
+			currentTick := mq.getTick()
+			if !window.alloc(result, currentTick) {
+				if bestEffort {
```

Comment:
This code doesn't return errors when it can't allocate quota; it returns 0 instead. Errors are used only for actual error conditions such as misconfiguration.

Regarding the semantics, it's the semantics are:

- I can support N units per seconds
- Please allocate M units right now, or as much as you can

I think this is a fairly straightforward model. It'll work well for prefetching tokens in particular.